### PR TITLE
broker resource secret authorization checking

### PR DESCRIPTION
**Pull request by jpeeler**
See #1167; ensures that the creator of a `ServiceBroker` resource is allowed to GET any secret that holds auth credentials for that broker.

### CODE REVIEW

**Comment by jpeeler** [comment]
Ok this is mostly completed now and is for issue #1167. I'm mostly wondering if it's worth checking on anything other than create (marked by TODO in the code).

File: cmd/apiserver/app/plugins.go [hunk 1]
```diff
@@ -21,6 +21,7 @@ package app
 // given binary target.
 import (
 	// Admission policies
```
Comment (pmorie) [inline cmd/apiserver/app/plugins.go]: Can you change this comment to say 'Admission controllers'

File: plugin/pkg/admission/broker/authsarcheck/admission.go [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+	if !s.WaitForReady() {
+		return admission.NewForbidden(a, fmt.Errorf("not yet ready to handle request"))
+	}
+	// only care about brokers
+	if a.GetResource().Group != servicecatalog.GroupName || a.GetResource().GroupResource() != servicecatalog.Resource("servicebrokers") {
+		return nil
+	}
+	broker, ok := a.GetObject().(*servicecatalog.ServiceBroker)
+	if !ok {
+		return errors.NewBadRequest("Resource was marked with kind Broker but was unable to be converted")
```
Comment (pmorie) [inline plugin/pkg/admission/broker/authsarcheck/admission.go]: Broker -> ServiceBroker

File: plugin/pkg/admission/broker/authsarcheck/admission.go [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+		return admission.NewForbidden(a, fmt.Errorf("broker does not have authentication information"))
+	}
+
+	var secretRef *v1.ObjectReference
+	if broker.Spec.AuthInfo.Basic != nil {
+		secretRef = broker.Spec.AuthInfo.Basic.SecretRef
+	} else if broker.Spec.AuthInfo.Bearer != nil {
+		secretRef = broker.Spec.AuthInfo.Bearer.SecretRef
+	} else if broker.Spec.AuthInfo.BasicAuthSecret != nil {
+		// this field is deprecated
```
Comment (pmorie) [inline plugin/pkg/admission/broker/authsarcheck/admission.go]: Leave a TODO here to remove this code when the field is removed in v1beta1 API.

File: plugin/pkg/admission/broker/authsarcheck/admission.go [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+// NewSARCheck creates a new namespace sarcheck admission control handler
+func NewSARCheck() (admission.Interface, error) {
+	return &sarcheck{
+		// TODO: are all these necessary?
+		Handler: admission.NewHandler(admission.Create, admission.Update, admission.Delete),
+	}, nil
+}
+
+func (s *sarcheck) SetKubeInformerFactory(f kubeinformers.SharedInformerFactory) {
+	namespaceInformer := f.Core().V1().Namespaces()
```
Comment (pmorie) [inline plugin/pkg/admission/broker/authsarcheck/admission.go]: why use the namespace informer?

File: plugin/pkg/admission/broker/authsarcheck/admission_test.go [hunk 5]
```diff
@@ -0,0 +1,158 @@
...
+							},
+						},
+					},
+				},
+			},
+			allowed: true,
+		},
+	}
+
+	userInfo := &user.DefaultInfo{
```
Comment (pmorie) [inline plugin/pkg/admission/broker/authsarcheck/admission_test.go]: why use the DefaultInfo type?

File: plugin/pkg/admission/broker/authsarcheck/admission_test.go [hunk 6]
```diff
@@ -0,0 +1,158 @@
...
+			name: "broker with no auth",
+			broker: &servicecatalog.ServiceBroker{
+				ObjectMeta: metav1.ObjectMeta{
+					Name: "test-broker",
+				},
+				Spec: servicecatalog.ServiceBrokerSpec{
+					URL: "http://example.com",
+				},
+			},
+			allowed: false,
```
Comment (pmorie) [inline plugin/pkg/admission/broker/authsarcheck/admission_test.go]: this should be allowed

File: plugin/pkg/admission/broker/authsarcheck/admission_test.go [hunk 7]
```diff
@@ -0,0 +1,158 @@
...
+								Name:      "test-secret",
+							},
+						},
+					},
+				},
+			},
+			allowed: true,
+		},
+	}
+
```
Comment (pmorie) [inline plugin/pkg/admission/broker/authsarcheck/admission_test.go]: we need a negative test where you pass an unexpected userinfo and get a negative result from the authorizer.

File: plugin/pkg/admission/broker/authsarcheck/admission_test.go [hunk 8]
```diff
@@ -0,0 +1,158 @@
...
+	pluginInitializer.Initialize(handler)
+	err = admission.Validate(handler)
+	return handler, kf, err
+}
+
+// newMockKubeClientForTest creates a mock kubernetes client that is configured
+// to allow any SAR creations.
+func newMockKubeClientForTest() *kubefake.Clientset {
+	mockClient := &kubefake.Clientset{}
+	mockClient.AddReactor("create", "subjectaccessreviews", func(action core.Action) (bool, runtime.Object, error) {
```
Comment (pmorie) [inline plugin/pkg/admission/broker/authsarcheck/admission_test.go]: make this only return allowed = true if you pass in a specific userinfo.

File: plugin/pkg/admission/broker/authsarcheck/admission.go [hunk 9]
```diff
@@ -0,0 +1,151 @@
...
+	}
+	glog.V(5).Infof("Retrieved broker %v", broker)
+
+	if broker.Spec.AuthInfo == nil {
+		// no auth secret to check
+		return admission.NewForbidden(a, fmt.Errorf("broker does not have authentication information"))
+	}
+
+	var secretRef *v1.ObjectReference
+	if broker.Spec.AuthInfo.Basic != nil {
```
Comment (nilebox) [inline plugin/pkg/admission/broker/authsarcheck/admission.go]: handle the case when SecretRef is not provided (no auth is configured in the broker, or the authInfo is invalid, missing secretRef)?
i.e. when secretRef is `nil`.

File: plugin/pkg/admission/broker/authsarcheck/admission.go [hunk 10]
```diff
@@ -0,0 +1,151 @@
...
+	}
+	broker, ok := a.GetObject().(*servicecatalog.ServiceBroker)
+	if !ok {
+		return errors.NewBadRequest("Resource was marked with kind Broker but was unable to be converted")
+	}
+	glog.V(5).Infof("Retrieved broker %v", broker)
+
+	if broker.Spec.AuthInfo == nil {
+		// no auth secret to check
+		return admission.NewForbidden(a, fmt.Errorf("broker does not have authentication information"))
```
Comment (nilebox) [inline plugin/pkg/admission/broker/authsarcheck/admission.go]: According to Service Catalog code ([broker.go](https://github.com/kubernetes-incubator/service-catalog/blob/master/pkg/apis/servicecatalog/validation/broker.go#L53)) and [OSB spec](https://github.com/openservicebrokerapi/servicebroker/blob/master/spec.md#authentication), authentication is optional.
So `AuthInfo == nil` is a valid case and shouldn't lead to errors.

File: plugin/pkg/admission/broker/authsarcheck/admission.go [hunk 11]
```diff
@@ -0,0 +1,151 @@
...
+		secretRef = broker.Spec.AuthInfo.Basic.SecretRef
+	} else if broker.Spec.AuthInfo.Bearer != nil {
+		secretRef = broker.Spec.AuthInfo.Bearer.SecretRef
+	} else if broker.Spec.AuthInfo.BasicAuthSecret != nil {
+		// this field is deprecated
+		secretRef = broker.Spec.AuthInfo.BasicAuthSecret
+	}
+	userInfo := a.GetUserInfo()
+
+	sar := &authorizationapi.SubjectAccessReview{
```
Comment (nilebox) [inline plugin/pkg/admission/broker/authsarcheck/admission.go]: handle the case when SecretRef above is not provided (the authInfo is invalid, missing secretRef)?
i.e. when secretRef is nil.

Comment (nilebox) [reply plugin/pkg/admission/broker/authsarcheck/admission.go]: @pmorie is validation executed before admission controller or after? if validation has succeeded at this point, then there is no need for further checking `secretRef != nil`

File: plugin/pkg/admission/broker/authsarcheck/admission.go [hunk 12]
```diff
@@ -0,0 +1,144 @@
...
+func (s *sarcheck) Admit(a admission.Attributes) error {
+	// need to wait for our caches to warm
+	if !s.WaitForReady() {
+		return admission.NewForbidden(a, fmt.Errorf("not yet ready to handle request"))
+	}
+	// only care about brokers
+	if a.GetResource().Group != servicecatalog.GroupName || a.GetResource().GroupResource() != servicecatalog.Resource("servicebrokers") {
+		return nil
+	}
+	serviceBroker, ok := a.GetObject().(*servicecatalog.ServiceBroker)
```
Comment (pmorie) [inline plugin/pkg/admission/broker/authsarcheck/admission.go]: You should check that the verb is create or update here.  I suspect what is failing in your PR tests is that you aren't getting the actual object from delete (IIRC you just will get the key of the object).

**Comment by pmorie** [comment]
@jpeeler 

Error from jenkins:

```
Deleting broker...
Error from server (BadRequest): Resource was marked with kind ServiceBroker, but was unable to be converted
/var/lib/jenkins/workspace/service-catalog-PR-testing2/src/github.com/kubernetes-incubator/service-catalog/contrib/jenkins/test_walkthrough.sh: line 212: Error when deleting ups-broker. (exit 1)
```

I suspect that what is happening is that you're not getting the object when you get a delete event.  Anyway, your admission controller should only operate when the verb is create or update.

**Comment by nilebox** [comment]
As mentioned in https://github.com/kubernetes-incubator/service-catalog/issues/1167#issuecomment-326855379, I'm not sure if admission controller is enough.
We can still merge this PR, but probably will need to extend it later.

File: plugin/pkg/admission/broker/authsarcheck/admission_test.go [hunk 13]
```diff
@@ -0,0 +1,192 @@
...
+				},
+			},
+			userInfo: &user.DefaultInfo{
+				Name:   "system:serviceaccount:test-ns:catalog",
+				Groups: []string{"system:serviceaccount", "system:serviceaccounts:test-ns"},
+			},
+			allowed: true,
+		},
+		{
+			name: "broker with basic auth, same namespace",
```
Comment (pmorie) [inline plugin/pkg/admission/broker/authsarcheck/admission_test.go]: brokers are not namespaced

File: plugin/pkg/admission/broker/authsarcheck/admission_test.go [hunk 14]
```diff
@@ -0,0 +1,192 @@
...
+				},
+			},
+			userInfo: &user.DefaultInfo{
+				Name:   "system:serviceaccount:test-ns:catalog",
+				Groups: []string{"system:serviceaccount", "system:serviceaccounts:test-ns"},
+			},
+			allowed: true,
+		},
+		{
+			name: "broker with bearer token, same namespace",
```
Comment (pmorie) [inline plugin/pkg/admission/broker/authsarcheck/admission_test.go]: brokers are not namespaced

File: plugin/pkg/admission/broker/authsarcheck/admission_test.go [hunk 15]
```diff
@@ -0,0 +1,192 @@
...
+	err = admission.Validate(handler)
+	return handler, kf, err
+}
+
+// newMockKubeClientForTest creates a mock kubernetes client that is configured
+// to allow any SAR creations.
+func newMockKubeClientForTest(userInfo *user.DefaultInfo) *kubefake.Clientset {
+	mockClient := &kubefake.Clientset{}
+	allowed := true
+	if userInfo.GetName() == "system:serviceaccount:test-ns:dave" {
```
Comment (pmorie) [inline plugin/pkg/admission/broker/authsarcheck/admission_test.go]: s/dave/forbidden/

**Comment by pmorie** [comment]
@jpeeler it looks like unit tests failed:

```
--- FAIL: TestAdmissionBroker (0.00s)

	admission_test.go:187: Create test 'broker with bearer token, same namespace, unauthenticated user' reports: Unexpected error returned from admission handler: <nil>

FAIL

FAIL	github.com/kubernetes-incubator/service-catalog/plugin/pkg/admission/broker/authsarcheck	1.284s
```

File: plugin/pkg/admission/broker/authsarcheck/admission_test.go [hunk 16]
```diff
@@ -0,0 +1,189 @@
...
+	for _, tc := range cases {
+		mockKubeClient := newMockKubeClientForTest(tc.userInfo)
+		handler, kubeInformerFactory, err := newHandlerForTest(mockKubeClient)
+		if err != nil {
+			t.Errorf("unexpected error initializing handler: %v", err)
+		}
+		kubeInformerFactory.Start(wait.NeverStop)
+
+		err = handler.Admit(admission.NewAttributesRecord(tc.broker, nil, servicecatalog.Kind("ServiceBroker").WithVersion("version"), tc.broker.Namespace, tc.broker.Name, servicecatalog.Resource("servicebrokers").WithVersion("version"), "", admission.Create, tc.userInfo))
+		if err != nil && tc.allowed == true || err == nil && tc.allowed == false {
```
Comment (kibbles-n-bytes) [inline plugin/pkg/admission/broker/authsarcheck/admission_test.go]: `tc.allowed == true` -> `tc.allowed`
`tc.allowed == false` -> `!tc.allowed`

File: plugin/pkg/admission/broker/authsarcheck/admission.go [hunk 17]
```diff
@@ -0,0 +1,143 @@
...
+		secretRef = serviceBroker.Spec.AuthInfo.BasicAuthSecret
+	}
+	userInfo := a.GetUserInfo()
+
+	sar := &authorizationapi.SubjectAccessReview{
+		Spec: authorizationapi.SubjectAccessReviewSpec{
+			ResourceAttributes: &authorizationapi.ResourceAttributes{
+				Namespace: secretRef.Namespace,
+				Verb:      "get",
+				Group:     "", // core API does not have a group
```
Comment (nilebox) [inline plugin/pkg/admission/broker/authsarcheck/admission.go]: `apiv1.SchemeGroupVersion.Group`
see https://github.com/kubernetes-incubator/service-catalog/blob/master/vendor/k8s.io/client-go/pkg/api/v1/register.go#L29

File: plugin/pkg/admission/broker/authsarcheck/admission.go [hunk 18]
```diff
@@ -0,0 +1,143 @@
...
+				Version:   "v1",
```
Comment (nilebox) [inline plugin/pkg/admission/broker/authsarcheck/admission.go]: `apiv1.SchemeGroupVersion.Version`
see https://github.com/kubernetes-incubator/service-catalog/blob/master/vendor/k8s.io/client-go/pkg/api/v1/register.go#L29

File: plugin/pkg/admission/broker/authsarcheck/admission.go [hunk 19]
```diff
@@ -0,0 +1,143 @@
...
+				Resource:  "secrets",
```
Comment (nilebox) [inline plugin/pkg/admission/broker/authsarcheck/admission.go]: You can reference `apiv1.ResourceSecrets.String()` constant here for resource name if you like.
https://github.com/kubernetes-incubator/service-catalog/blob/master/vendor/k8s.io/client-go/pkg/api/v1/types.go#L4140

File: plugin/pkg/admission/broker/authsarcheck/admission.go [hunk 20]
```diff
@@ -0,0 +1,143 @@
...
+				Name:      secretRef.Name,
+			},
+			User:   userInfo.GetName(),
+			Groups: userInfo.GetGroups(),
+			Extra:  convertToSARExtra(userInfo.GetExtra()),
+			// uncomment below next rebase
+			//UID:    userInfo.GetUID(),
```
Comment (nilebox) [inline plugin/pkg/admission/broker/authsarcheck/admission.go]: ?

File: plugin/pkg/admission/broker/authsarcheck/admission_test.go [hunk 21]
```diff
@@ -0,0 +1,189 @@
...
+		{
+			name: "broker with basic auth, user authenticated",
+			broker: &servicecatalog.ServiceBroker{
+				ObjectMeta: metav1.ObjectMeta{
+					Name: "test-broker",
+				},
+				Spec: servicecatalog.ServiceBrokerSpec{
+					URL: "http://example.com",
+					AuthInfo: &servicecatalog.ServiceBrokerAuthInfo{
+						Basic: &servicecatalog.BasicAuthConfig{
```
Comment (nilebox) [inline plugin/pkg/admission/broker/authsarcheck/admission_test.go]: Can you add a test for the same use case but with deprecated `serviceBroker.Spec.AuthInfo.BasicAuthSecret` field set instead?

File: plugin/pkg/admission/broker/authsarcheck/admission.go (hunk 20 above)
Comment (jpeeler) [reply plugin/pkg/admission/broker/authsarcheck/admission.go]: It's coming eventually: https://github.com/kubernetes/kubernetes/blob/f14c1384387ac196e87334b5a0e05e01d7581387/pkg/apis/authorization/types.go#L125

File: plugin/pkg/admission/broker/authsarcheck/admission_test.go (hunk 21 above)
Comment (jpeeler) [reply plugin/pkg/admission/broker/authsarcheck/admission_test.go]: added!

File: plugin/pkg/admission/broker/authsarcheck/admission.go (hunk 20 above)
Comment (nilebox) [reply plugin/pkg/admission/broker/authsarcheck/admission.go]: @jpeeler can you leave a more specific TODO comment then? Something like:
```
// TODO uncomment after rebase onto Kubernetes 1.8.
// See https://github.com/kubernetes/kubernetes/pull/49677
```

**Comment by nilebox** [comment]
@jpeeler good job, please fix the TODO for UID and I will LGTM and merge it.

**Comment by kibbles-n-bytes** [comment]
@nilebox TODO has been fixed, I'm going to go ahead and LGTM and merge!
